### PR TITLE
[BEAM-2203] Implement TIMESTAMPADD

### PR DESCRIPTION
**Pull request by akedin**
Add support for TIMESTAMPADD(interval, multiplier, TIMESTAMP)

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by akedin** [comment]
R: @XuMingmin  @xumingming 
Hi, could you please take a look?

**Comment by kennknowles** [comment]
Looks like Jenkins works `beam8` is messed up, and is getting a lot of jobs. I'll contact infra. Meanwhile, I'll ask Jenkins to run these again.

**Comment by kennknowles** [comment]
retest this please

File: sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/integrationtest/BeamSqlDateFunctionsIntegrationTest.java [hunk 1]
```diff
@@ -50,7 +51,9 @@
         .addExpr("SECOND(ts)", 26L)
         .addExpr("FLOOR(ts TO YEAR)", parseDate("1986-01-01 00:00:00"))
         .addExpr("CEIL(ts TO YEAR)", parseDate("1987-01-01 00:00:00"))
-        ;
+        .addExpr("TIMESTAMPADD(DAY, 3, TIMESTAMP '1984-04-19 01:02:03')",
```
Comment (XuMingmin) [inline sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/integrationtest/BeamSqlDateFunctionsIntegrationTest.java]: can you add more elements to cover YAER, MONTH, DAT, HOUR, MINUTE, SECOND?

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/interpreter/operator/date/BeamSqlDatetimePlusExpression.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+import org.apache.beam.sdk.extensions.sql.impl.interpreter.operator.BeamSqlExpression;
+import org.apache.beam.sdk.extensions.sql.impl.interpreter.operator.BeamSqlPrimitive;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.values.BeamRecord;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.joda.time.DateTime;
+
+/**
+ * DATETIME_PLUS operation.
```
Comment (XuMingmin) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/interpreter/operator/date/BeamSqlDatetimePlusExpression.java]: please add the expression returned from Calcite parser, that would be helpful to know the difference between `TIMESTAMPADD` written in user SQL, and the actual execution expression in the backend.

Comment (akedin) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/interpreter/operator/date/BeamSqlDatetimePlusExpression.java]: `TIMESTAMPADD()` and infix `+` (more specifically `DATE + INTERVAL`) are converted by Calcite into `DATETIME_PLUS`. I'll update the comment and will add few more tests

File: sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/integrationtest/BeamSqlDateFunctionsIntegrationTest.java (hunk 1 above)
Comment (akedin) [reply sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/integrationtest/BeamSqlDateFunctionsIntegrationTest.java]: Sure, I'll add them here. Plus I will check if the infix version works correctly

**Pull request closed without merging**